// source-map-js/Mascot-svg-only 2/src/basic/Rectangle.js
export default class Rectangle {

	constructor(left, top, width, height) {
		this.left = left;
		this.top = top;
		this.width = width;
		this.height = height;
	}

	toJSON() {
		let json = {};
		json.left = this.left;
		json.top = this.top;
		json.width = this.width;
		json.height = this.height;
		return json;
	}

	union(rect) {
		let left = Math.min(this.left, rect.left),
			top = Math.min(this.top, rect.top),
			right = Math.max(this.right, rect.right),
			btm = Math.max(this.bottom, rect.bottom);
		return new Rectangle(left, top, right - left, btm - top);
	}

	clone() {
		return new Rectangle(this.left, this.top, this.width, this.height);
	}

	get right() {
		return this.left + this.width;
	}

	get bottom() {
		return this.top + this.height;
	}

	get x() {
		return (this.left + this.right)/2;
	}

	get y() {
		return (this.top + this.bottom)/2;
	}

	get center() {
		return (this.left + this.right)/2;
	}

	get middle() {
		return (this.top + this.bottom)/2;
	}

	contains(x, y) {
		return this.left <= x && this.right >= x && this.top <= y && this.bottom >= y;
	}

	overlap(r) {
		return !(this.right < r.left || this.bottom < r.top || this.left > r.right || this.top > r.bottom);
	}

}
