// mascot-svg-only/rollup.config.js
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import terser from "@rollup/plugin-terser";
import pkg from './package.json' assert { type: 'json' };


const external = [ 'three', 'd3', "pixi.js" ];
const globals = { 'three': 'THREE', "d3": "d3", "pixi.js" : "PIXI" };

export default [
	{
		input: "src/index.js",
		output: [
			{
				file: "dist/mascot.js",
				name: "msc",
				format: "umd",
				globals,
				banner: "/* eslint-disable */\n// version: " + pkg.version
			},
			{
				file: "dist/mascot-min.js",
				name: "msc",
				format: "umd",
				globals,
				banner: "/* version: " + pkg.version + " */",
				plugins: [terser({
					output: {
						comments: '/version(.*)/' 
					}
				})]
			}
		],
		external,
		plugins: [
			resolve({
				preferBuiltins: false
			}),			
			commonjs()
		],
		onwarn: function (warning, warn) {
			if (warning.code === 'CIRCULAR_DEPENDENCY') return;
			warn(warning);
		}
	},
	//exclude PixiRenderer as Pixi requires browser context
	{
		input: "src/indexSVG.js",
		output:
		{
			file: "dist/mascotSVG.js",
			name: "msc",
			format: "umd",
			globals: { "d3": "d3" }
		},
		external: ['d3'],
		plugins: [
			resolve({
				preferBuiltins: false
			}),
			commonjs()
		],
		onwarn: function (warning, warn) {
			if (warning.code === 'CIRCULAR_DEPENDENCY') return;
			warn(warning);
		}
	}
];
